### DISCUSSION TITLE
How to properly use `beartype` for a single internal module of a package?

### DISCUSSION THREAD

**Discussion by taranlu-houzz** [beartype/beartype#434] [Q&A]
My project has the following structure:
```
.
├── src
│  └── A
│     └── B
│        └── C
│           └── D
│              ├── _private
│              │  ├── exceptions.py
│              │  ├── lib.py
│              │  └── logging.py
│              └── __init__.py
├── pyproject.toml
└── uv.lock
```

When I try to use `beartype.claw.beartype_this_package()` or `beartype.claw.beartype_package(__name__)` in `lib`, it doesn't seem to work. However, I am able to use the `beartype.beartype` decorator directly. The reason I am not using `beartype` in the package `__init__` is because the actual project uses `typer`, which does not play nicely with `beartype`, so I just want to use strict typing for the internal library functions for now.

**Comment by leycec** [comment]
**Excellent questions!** Happily, I can actually solve everything for once. Let's do this.

> When I try to use `beartype.claw.beartype_this_package()` or `beartype.claw.beartype_package(__name__)` in `lib`, it doesn't seem to work.

**Right.** Sadly, that's exactly what one would expect. I know. Kinda lame, isn't it?

Allow me to now pontificate. Like all Python import hooks, `beartype.claw` import hooks (like `beartype_this_package()` and `beartype_package()`) only apply to *subsequent* imports. Crucially, `beartype.claw` import hooks do *not* apply to the current module (like your `A.B.C.D._private.lib` submodule), as the current module has already been imported. *Makes sense, sorta.*

The usual solution is thus to call `beartype.claw` import hooks from the `__init__` submodule of a package. Since Python implicitly imports the `__init__` submodule of a package *before* importing anything else from that package, calling `beartype.claw` import hooks from the `__init__` submodule guarantees type-checking across all submodules of that package. *Makes sense, sorta.*

## tl;dr: Five-Second Solutions Feels Good

Please call `beartype_package()` from *any* `__init__` submodule at or above the level of your `A.B.C.D._private` subpackage. I only see one `__init__` submodule listed above: the `A.B.C.D.__init__` submodule. You *probably* have a lot more `__init__` submodules lying around, though. Any of the following would also work:

* The `A.B.C.D._private.__init__` submodule.
* The `A.B.C.D.__init__` submodule.
* The `A.B.C.__init__` submodule.
* The `A.B.__init__` submodule.
* The `A.__init__` submodule.

Whichever you prefer! Any should work. Specifically, this is what I advise:

```python
# In any of the above "__init__" submodules:
from beartype.claw import beartype_package
beartype_package('A.B.C.D._private.lib')
```

Oh – and note that `beartype_package()` requires **absolute package names.** That's why I've laboriously spelled out the absolute `'A.B.C.D._private.lib'` subpackage name above. *Boring stuff.*

> The reason I am not using beartype in the package `__init__` is because the actual project uses `typer`, which does not play nicely with beartype...

**Sadness!** I actually didn't know that. Ideally, @beartype should play nicely with everybody. @beartype strictly adheres to *almost* all PEP (Python Enhancement Proposal) standards. If @beartype fails to play nicely with `typer`, this might be a `typer` issue. Alternately, @beartype might be subtly broken in some way. In either case, this is why we can't have good CLI QA. :face_exhaling: 

If you find the time to produce a minimal-reproducible example (MRE) and a traceback, would you mind submitting a new issue to the @beartype issue tracker about this integration issue between @beartype + `typer`? @beartype and I both thank you so much in advance. :pray:

**Comment by taranlu-houzz** [comment]
Hey @leycec, thank you for the thorough response! I will try out the solution you mentioned about using `beartype_package()` in one of the `__init__.py` modules. I actually tend to try to keep the empty `__init__.py` modules out of my projects unless absolutely necessary, so the structure I have you is actually how my project is (e.g. no `__init__.py` for the internal `_private` dir. I'll try adding that in and using your proposed solution when I have time to circle back to this.

As far as the `typer` issue goes, it mainly has to do with how one turns normal functions into `typer` commands. Since `typer` also does validation and some other complex work based on your function params, it is often necessary to use the `typer.Option` or `typer.Argument` objects, which triggers `beartype` if you try to use the function (that had been decorated by `typer`) with that actual expected values that would be passed to the function.

```python
@APP.command()
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```
So, for example, the `submit_change` param expects a `typer.Option` object as far as `beartype` is concerned, even though it would be reasonable to call this function by passing a `bool` directly.

**Reply by leycec** [reply]
**Ah!** Got it. Thanks so much for the detailed example. Given that, it should be no trouble whatsoever for @beartype to officially support `typer`. I'll just need a spare moment of free time and... *that's never gonna happen, is it*!?

Maybe not. I really hope it does, though! `typer` is nifty. Nifty packages deserve our support. So, I've converted your awesome commentary into a new feature request #436.

## What Can Be Done Until @leycec Does Something, Though?

Thankfully, you don't have to wait for @leycec to finally do something. Instead, just tell @beartype that it should ignore all `@APP.command()`-decorated callables like so:

```python
from typing import no_type_check

@APP.command()
@no_type_check
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```

Like all runtime type-checkers (e.g., `typeguard`) and static type-checkers (e.g., `mypy`, `pyright`), @beartype silently ignores *all* callables and classes decorated by the standard `@typing.no_type_check` decorator. Since `typer` kinda violates PEP standards, <sup>*...sad!*</sup> it's helpful to notify type-checkers that they should ignore `typer`-based callables.

Decorating `typer`-based callables with `@typing.no_type_check` then allows you to use the convenient `beartype.claw.beartype_this_package()` import hook to globally runtime type-check your entire package. This is how the QA was won.

@beartype + `typer`: *soon to be a match made in CLI heaven.* :angel:

**Comment by taranlu-houzz** [comment]
@leycec I'm probably still doing something wrong, but when I tried adding this `__init__.py` to my internal `_private` dir to make it a package, I am not getting any effect on the functions in the `lib` module.

```python
"""Private internal package."""

# ==Site-Packages==
from beartype.claw import beartype_package


####################################################################################################
# Main
####################################################################################################


beartype_package(f"{__package__}.lib")
```
I also did check with a breakpoint right before the `beartype_package` call, and the value of `__package__` is what I would expect (e.g. `A.B.C.D._private`).

Is there a way to verify what hooks were registered and what functions/modules are affected?

**Reply by leycec** [reply]
**Gah!** Sadly...

> Is there a way to verify what hooks were registered and what functions/modules are affected?

Tragically, no such public API exists at the moment. This is the first case in which `beartype.claw` isn't doing what someone tells `beartype.claw` to do – to my knowledge, anyway. Clearly, this calls for a change. Let's open up a [new feature request tracking this *extremely* useful feature request](https://github.com/beartype/beartype/issues/437).

Superficially, it *looks* like you're doing everything right...

## WAIT. I Stupidly Glossed Over Your Comment Earlier!

I feel growing fear for your code, @taranlu-houzz:

> I actually tend to try to keep the empty `__init__.py` modules out of my projects unless absolutely necessary, so the structure I have you is actually how my project is (e.g. no `__init__.py` for the internal `_private` dir).

**Ho, ho, ho.** You're now swimming with sharks in dangerous waters. I see that you like to live fast and loose. I also understand your *totally* valid concerns. From the post-modern 2024 perspective, empty `__init__.py` modules seem really weird, superfluous, and an archaic holdover from the violent medieval era of Python 2.7. You are now thinking: "Who needs those silly things anyway?"

Sadly, the answer is: "Everybody. Everybody needs those silly things." None of us like empty `__init__.py` modules, but we all still need those silly things. **Empty `__init__.py` modules are absolutely required at all packaging levels throughout every single subdirectory of your package.** This means that your package structure *really* does need to resemble:

```
.
├── src
│  └── A
│     ├── __init__.py
│     └── B
│        ├── __init__.py
│        └── C
│           ├── __init__.py
│           └── D
│              ├── __init__.py
│              └── _private
│                 ├── __init__.py
│                 ├── exceptions.py
│                 ├── lib.py
│                 └── logging.py
├── pyproject.toml
└── uv.lock
```

Love that ASCII art, by the way. So h0t! :heart_eyes: 

## But *Why* Do I Need Those Awful Empty `__init__.py` Modules? WHY!?!?

*...heh.* I feel you. The reason is simple. Actually, the reason is complex.

**Python implicitly treats *any* package subdirectory lacking an `__init__.py` module as an "implicit namespace package."** If you don't know *exactly* what an "implicit namespace package" is, you definitely shouldn't be creating them – which means you *really* shouldn't be omitting `__init__.py` modules. Implicit namespace packages are fundamentally dangerous. They're *not* something you want to do unless you really know that you need to do that thing, despite the shark-infested waters.

Implicit namespace packages create a *heap* of trouble for everybody in the Python space. They're actually quite rare. I've only ever encountered a single open-source namespace package in the wild: [`ruamel.yaml`](https://pypi.org/project/ruamel.yaml), Python's best YAML (and thus JSON) parser by a wide margin. The `ruamel` in `ruamel.yaml` is a namespace package; meanwhile, the `yaml` in `ruamel.yaml` is a normal non-namespace subpackage.

As far as I know, static type-checkers like `mypy` and `pyright` *still* have difficulty with implicit namespace packages. Most Python tools and IDEs do. They're weird; they don't import or behave correctly; they're not "real" packages in the normal sense of "real."

But... maybe you *do* intentionally want your `A`, `A.B`, `A.B.C`, and `A.B.C.D` folders to all be treated as implicit namespace packages? This would be the case if, for example, the funding for your package is governed by an `A` organization with a `B` suborganization and so on – and all of these organizations publish a variety of Python packages controlled by different teams and devs. But... *yeah.* That's probably not the case. Is it? No idea. I'm just a bald dude on GitHub. :rofl:

## What Do Namespace Packages Have to Do with Import Hooks?

No idea. I don't think anyone's attempted to apply @beartype import hooks to namespace packages before. You're the first – and it doesn't seem to be going well. It's possible that namespace packages aren't hookable by name. Maybe? Seems plausible. Namespace packages are weird. Unfortunately, I never tested that. Awkward moment, huh? In my defence, namespace packages are so rare "in the wild" that I *never* expected anyone to actually create one – let alone apply @beartype to one.

Guess I should've probably tested that. :face_exhaling: 

## One Last Idea: Does `beartype_this_package()` + `@no_type_check` Work?

As a last-ditch desperate idea, would you mind trying to:

* Additionally decorate all of your `typer`-specific callables with `@typing.no_type_check`?
* Replacing that call to `beartype_package(f"{__package__}.lib")` with just `beartype_this_package()`? This should *always* work: e.g.,

```python
# In your "A.B.C.D._private.__init__" submodule:
from beartype.claw import beartype_this_package
beartype_this_package()
```

If even that fails, there *is* a secret `print()` statement in the @beartype codebase that debugs exactly what you want to debug. Sadly, it's commented out; you'll have to manually uncomment it. The `print()` statement in question lives in the `beartype/claw/_importlib/_clawimpload.py` submodule. Just:

```python
# In the "beartype/claw/_importlib/_clawimpload.py" submodule,
# replace this commented-out code black:
        #FIXME: Conditionally perform this logic if "conf.is_debug", please.
        #Note that printing to "stderr" is pivotal. For some reason, Python
        #fails to forward printing to "stdout" across subprocesses even when we
        #explicitly tell it to. Look. I don't even know. Just roll with it!
        # from sys import stderr
        # print(
        #     (
        #         f'Module "{self._module_name_beartype}" abstract syntax tree (AST) '
        #         f'transformed by @beartype to:\n\n'
        #         f'{get_node_repr_indented(module_ast_beartyped)}'
        #     ),
        #     file=stderr,
        # )

# ...with this uncommented code block:
        from sys import stderr
        print(
            (
                f'Module "{self._module_name_beartype}" abstract syntax tree (AST) '
                f'transformed by @beartype to:\n\n'
                f'{get_node_repr_indented(module_ast_beartyped)}'
            ),
            file=stderr,
        )
```

So sorry about that! We really do need that new `BeartypeConf(claw_is_trace_imports)` option. :cry:

**Comment by taranlu-houzz** [comment]
@leycec Wow, that is quite the response! Thank you! Here is some more background on my use case:
- I know that the lack of explicit `__init__.py` is used to define implicit namespace packages; that is actually my intention and is pretty much how I manage all of my projects.
- I like namespace packages for the following reasons:
  - Able to categorize (and sub-categorize) my different related code/tools:
    - My actual structure is like this: `<company>.<team/repo>.<category>.<package name>`
      - `my_company.pipeline.tools.my_cool_cli_app`
      - `my_company.pipeline.servers.my_special_server_app`
  - It is also a bit of a pain to bug IT to get help creating new repos within the org, so having this naming scheme allows me to have a nice clean mono-repo structure like this:
    ```
    .
    └── src
       └── python
          ├── my_company.pipeline.legacy.libs.cool_lib
          ├── my_company.pipeline.legacy.tools.cool_tool
          ├── my_company.pipeline.libs.even_cooler_lib
          ├── my_company.pipeline.servers.cool_server
          ├── my_company.pipeline.tools.even_cooler_tool
          └── my_company.pipeline.tools.another_cool_tool
    ```
- I have run into some minor issues with namespace packages over the years, but I don't think many (any?) of them were due to them being implicit namespace packages.
- At least for me, my current IDE-ish setup with VSCode seems to work reasonably well.
- My experience with `mypy` has been mixed, but I do tend to enable it with the following setting on most projects I start now:
  ```toml
  [tool.mypy]
  ignore_missing_imports = true
  mypy_path = "$MYPY_CONFIG_FILE_DIR/src:$MYPY_CONFIG_FILE_DIR/tests"
  namespace_packages = true
  strict = true
  ```

I will try out what you suggested in the `One Last Idea` section. At first, when I did that without using the `@no_type_check` decorator, it immediately caused `beartype` to complain, which is kind of how this all started.

Really appreciate your thorough responses!

**Comment by leycec** [comment]
**WOAH.** You... you're doing everything right! I'm astonished, actually. Namespace packages are kinda esoteric. Your use case is ideal and perfect for namespace packages. My assumption was *waaaay* off. I humbly apologize for insinuating that your package structure was inappropriate or somehow "bad." Indeed, I'm bad. The design pattern you've adhered to is a flawless example of namespace packaging. Which leads us to only one valid conclusion, really:

> `beartype.claw` import hooks are *probably* incompatible with namespace packaging – *probably.*

Namespace packaging is *really* non-trivial. Usually, that's not an issue. In the case of import hooks, though, that's an issue. All import hooks (including those published through the `beartype.claw` API) are implemented in terms of [`importlib` machinery](https://docs.python.org/3/library/importlib.html). `importlib` is the *extremely* low-level esoteric standard CPython API that governs `import` statements. It's *incredibly* obtuse and difficult to fully "grok." I am convinced that no one (including Guido himself) completely understands the modern implementation of `importlib`.

Namespace packages are fundamentally different from standard packages. Thus, so is their importation. Thus, so is their implementation via the `importlib` API. Thus, so is their usage from import hooks like `beartype.claw`. The way that `beartype.claw` "hooks" namespace packages *probably* needs to fundamentally differ from the way that `beartype.claw` "hooks" standard packages. In other words, `beartype.claw` needs explicit support for namespace packages. In the absence of explicit support, it is unsurprising that `beartype.claw` fails to support namespace packages. I weep openly.

Let's open yet another feature request #438 for this! Someday, @beartype will actually work for you, @taranlu-houzz. Until then, I beg for*bear*ance. <sup>*...get it? for*bear*ance? i'll stop now.*</sup>

Forgive @beartype its sins. It knows not what it does. :face_holding_back_tears:

**Comment by taranlu-houzz** [comment]
@leycec Haha, thanks for the info. Interesting to hear that namespace packages are such an issue. That is unfortunate. As a followup, I did try using `@typing.no_type_check` and it unfortunately did not work (at least, as far as preventing `beartype` from triggering goes). It also seems like it would prevent `mypy` from checking those functions, which isn't quite ideal.

Interestingly, `mypy` was sometimes(?) picky about the order that decorator gets applied; it seemed to want it to be at the top of the stack (but then sometimes it would not highlight that as an issue):
```python
@typing.no_type_check
@APP.command() # typer decorator
def my_func():
    ...
```

I think that for now, I will probably just directly decorate the functions with `beartype.beartype` just to make things predictable.

I did run into one more interesting issue though. When I ran my tool via its cli command (as opposed to using the lib part), `beartype` did catch a runtime type violation (good!!), but the error message that it logs gets messed up by the pretty exception handling that `typer` does using the `rich` library. It seems to output the message with `rich` markup codes embedded:
```
BeartypeDoorHintViolation: Callable [1m[35m----._private.lib.save_tile_data()[0m local variable "[1m[35mmetadata[0m" value [1m[31m"None"[0m
violates type hint [1m[32mdict[str, typing.Any][0m, as [1m[33m<class "builtins.NoneType"> [0m[1m[31m"None"[0m not instance of [1m[32mdict[0m.
```
It would be nice to have official support for pretty `rich` exception logging integration for `beartype` (maybe as an optional dependency?).

Edit: Here is a link to the `typer` docs that show the nice `rich` exceptions: https://typer.tiangolo.com/tutorial/exceptions/?h=rich+tr#exceptions-with-rich

Edit: I tried using the `beartype.beartype` decorator, and it does work (I was able to trigger it by giving the func args with the wrong type). Strangely, it now does not error at the section inside the function the way it did when I used `beartype.claw` to `beartype` the whole package. Seems like that must be doing something different/deeper?

**Reply by leycec** [reply]
> I did try using `@typing.no_type_check` and it unfortunately did not work (at least, as far as preventing `beartype` from triggering goes).

Madness continues, huh? My tired face is tired, fam. :face_exhaling: 

> It also seems like it would prevent `mypy` from checking those functions, which isn't quite ideal.

*...heh.* You noticed even that, huh? Nothing slips past @taranlu-houzz. I didn't want to muddy the waters any further, but the @beartype API *does* actually provide a @beartype-specific means of selectively **blacklisting** (i.e., ignoring) undesirable callables and classes: the `O(0)` type-checking strategy, meaning "do literally nothing by silently reducing to a noop, @beartype." Here's what that looks like for this specific use case:

```python
from beartype import beartype, BeartypeConf, BeartypeStrategy

# "beartype.claw" import hooks will now ignore this sync() function. Yay!
@APP.command()
@beartype(conf=BeartypeConf(strategy=BeartypeStrategy.O0))
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```

If you get exhausted of copy-pasting `@beartype(conf=BeartypeConf(strategy=BeartypeStrategy.O0))` everywhere, consider aliasing that to a private global ala:

```python
from beartype import beartype, BeartypeConf, BeartypeStrategy

_no_beartype = beartype(conf=BeartypeConf(strategy=BeartypeStrategy.O0))
'''
:func:`beartype.beartype` decorator configured to do absolutely nuffin'.
'''

@APP.command()
@_no_beartype
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```

If that *still* doesn't work, I will eat my priceless Tilly hat live on GitHub. :womans_hat: :yum: 

> ...but the error message that it logs gets messed up by the pretty exception handling that `typer` does using the rich library. It seems to output the message with `rich` markup codes embedded:

**Ah-ha!** Those are [Good Ol' ANSI Escape Sequences™ (GOAES)](https://en.wikipedia.org/wiki/ANSI_escape_code), actually. Substrings like `[0m` and `[1m[32m` are the telltale giveaway. They were used prominently during the 1980's [Bulletin Board System (BBS)](https://en.wikipedia.org/wiki/Bulletin_board_system) craze, but are now just a relic of a long-forgotten age that nobody wants to think about anymore. This is almost certainly an integration issue between @beartype and whatever logging system you're currently using. Fear not! An old bald guy named @leycec will now solve all your problems.

By default, @beartype colourizes **type-checking violations** (i.e., `beartype.roar.BeartypeViolation` exceptions) by embedding Good Ol' ANSI Escape Sequences™ (GOAES) in violation messages. Thankfully, you can disable that via a similar mechanism as above: by passing the `conf=BeartypeConf` parameter to either the `@beartype` decorator *or* a `beartype.claw` import hook.

If you currently prefer the `@beartype` decorator, I totally get you. One human can only struggle with `beartype.claw` so much before giving up. In this case, you'll *probably* want to create your own `@beartype` decorator uniquely configured for your codebase: e.g.,

```python
# Declare this somewhere useful in your codebase -- like, say, a
# "A.B.C.D._private.util" submodule or something similar.
from beartype import beartype, BeartypeConf

monobeartype = beartype(conf=BeartypeConf(is_color=False))
'''
:func:`beartype.beartype` decorator configured to disable colourization.
'''
```

Then replace *all* uses of the standard `@beartype` decorator throughout your code with this new `@monobeartype` decorator: e.g.,

```python
# In some other submodule:
from A.B.C.D._private.util import monobeartype

@monobeartype
def awesome_func(so_much_awesome: list[str]) -> int: ...
```

Unsurprisingly, this is one of many face-palming reasons why the `beartype.claw` API is dramatically superior – assuming it actually worked, anyway! Rather than having to globally refactor your usage of `@beartype` (*...pretty annoying, in all honesty*), you'd just revise your call of a `beartype.claw` import hook to resemble:

```python
# In some "...__init__" submodule somewhere:
from beartype import beartype, BeartypeConf
from beartype.claw import beartype_this_package
beartype_this_package(conf=BeartypeConf(is_color=False))
```

> Strangely, it now does not error at the section inside the function the way it did when I used `beartype.claw` to `beartype` the whole package. Seems like that must be doing something different/deeper?

**Ho, ho, ho.** Madness intensifies. Indeed, you have discovered the truth: @beartype is a hybrid runtime-static type-checker. This means that @beartype performs both static type-checking (like `mypy` and `pyright`) *and* runtime type-checking (like, uhh, @beartype and `typeguard`). To minimize false positives and maximize correctness, @beartype prefers to do most of its type-checking at runtime. Unfortunately, some forms of type-checking can *only* be performed statically. This includes [PEP 526-compliant annotated variable assignments](https://peps.python.org/pep-0526/) like:

```python
# @beartype only type-checks annotated variable assignments like this when
# @beartype is applied via a "beartype.claw" import hook. Sadness and confusion ensues.
some_global: list[str] = [42, 3.14]  # <--- *BUG*
```

As the above comment suggests, @beartype **only performs static type-checking when applied via a `beartype.claw` import hook.** The lower-level `@beartype.beartype` decorator is incapable of type-checking [PEP 526-compliant annotated variable assignments](https://peps.python.org/pep-0526/) like above; *all* of those assignments will be silently ignored unless a `beartype.claw` import hook is run.

I swear this isn't our fault! It's kinda just how Python works. I blame Guido for everything. :face_holding_back_tears:
